From e26b6b177aaaaf4369d90c941d67bcd0a7321436 Mon Sep 17 00:00:00 2001
From: Hansong Zhang <hsz@google.com>
Date: Wed, 10 Feb 2021 14:39:33 -0800
Subject: [PATCH] Check packet size in avrc_pars_browse_rsp

Test: manual
Bug: 179162665
Change-Id: If4d15b3fe907055f3691d03467a3a64dc315fa71
Tag: #security
(cherry picked from commit 6e3c984806aa0ba9e8a836ef9fed12c1207a161e)
---
 stack/avrc/avrc_pars_ct.cc | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/stack/avrc/avrc_pars_ct.cc b/stack/avrc/avrc_pars_ct.cc
index 7c6367644..08ab66e72 100644
--- a/stack/avrc/avrc_pars_ct.cc
+++ b/stack/avrc/avrc_pars_ct.cc
@@ -428,12 +428,19 @@ static tAVRC_STS avrc_pars_browse_rsp(tAVRC_MSG_BROWSE* p_msg,
     case AVRC_PDU_GET_ITEM_ATTRIBUTES: {
       tAVRC_GET_ATTRS_RSP* get_attr_rsp = &(p_rsp->get_attrs);
       get_attr_rsp->pdu = pdu;
+      min_len += 2;
+      if (pkt_len < min_len) {
+        android_errorWriteLog(0x534e4554, "179162665");
+        goto browse_length_error;
+      }
       BE_STREAM_TO_UINT8(get_attr_rsp->status, p)
       BE_STREAM_TO_UINT8(get_attr_rsp->num_attrs, p);
       get_attr_rsp->p_attrs = (tAVRC_ATTR_ENTRY*)osi_malloc(
           get_attr_rsp->num_attrs * sizeof(tAVRC_ATTR_ENTRY));
       for (int i = 0; i < get_attr_rsp->num_attrs; i++) {
         tAVRC_ATTR_ENTRY* attr_entry = &(get_attr_rsp->p_attrs[i]);
+        min_len += 8;
+        if (pkt_len < min_len) goto browse_length_error;
         BE_STREAM_TO_UINT32(attr_entry->attr_id, p);
         BE_STREAM_TO_UINT16(attr_entry->name.charset_id, p);
         BE_STREAM_TO_UINT16(attr_entry->name.str_len, p);
--
2.29.0
